device/core/src/sak_authentication_provider.ts

```diff
@@ -0,0 +1,130 @@
+  }
+
+  private _renewToken(): void {
+    if (this._renewalTimeout) {
+      clearTimeout(this._renewalTimeout);
+    }
+
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_009: [Every token shall be created with a validity period of `tokenValidTimeInSeconds` if specified when the constructor was called, or 1 hour by default.]*/
+    const newExpiry =  Math.floor(Date.now() / 1000) + this._tokenValidTimeInSeconds;
+
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_010: [Every token shall be created using the `azure-iot-common.SharedAccessSignature.create` method and then serialized as a string, with the arguments to the create methods being:
+    ```
+    resourceUri: <IoT hub host>/devices/<deviceId>
+    keyName: the `SharedAccessKeyName` parameter of the connection string or `null`
+    key: the `SharedAccessKey` parameter of the connection string
+    expiry: the expiration time of the token, which is now + the token validity time, formatted as the number of seconds since Epoch (Jan 1st, 1970, 00:00 UTC).
+    ```]*/
+    const resourceUri = encodeUriComponentStrict(this._credentials.host + '/devices/' + this._credentials.deviceId);
+    const sas = SharedAccessSignature.create(resourceUri, this._credentials.sharedAccessKeyName, this._credentials.sharedAccessKey, newExpiry);
+    this._credentials.sharedAccessSignature = sas.toString();
+
+    const nextRenewalTimeout = (this._tokenValidTimeInSeconds - this._tokenRenewalMarginInSeconds) * 1000;
+    if (this.automaticRenewal) {
+      this._renewalTimeout = setTimeout(() => this._renewToken(), nextRenewalTimeout);
+    }
+    /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_005: [Every time a new token is created, the `newTokenAvailable` event shall be fired with no arguments.]*/
+    this.emit('newTokenAvailable');
+  }
+
+  /**
+   * Creates a new `SharedAccessKeyAuthenticationProvider` from a connection string
+   *
+   * @param connectionString              A device connection string containing the required parameters for authentication with the IoT hub.
+   * @param tokenValidTimeInSeconds       [optional] The number of seconds for which a token is supposed to be valid.
+   * @param tokenRenewalMarginInSeconds   [optional] The number of seconds before the end of the validity period during which the `SharedAccessKeyAuthenticationProvider` should renew the token.
+   */
+  static fromConnectionString(connectionString: string, tokenValidTimeInSeconds?: number, tokenRenewalMarginInSeconds?: number): SharedAccessKeyAuthenticationProvider {
+    if (!connectionString) {
+      /*Codes_SRS_NODE_SAK_AUTH_PROVIDER_16_006: [The `fromConnectionString` method shall throw a `ReferenceError` if the `connectionString` parameter is falsy.]*/
+      throw new ReferenceError('connectionString cannot be \'' + connectionString + '\'');
+    }
+
+    const cs = ConnectionString.parse(connectionString);
+    if (!cs.SharedAccessKey) {
```

Comment:
>SharedAccessKey [](start = 12, length = 15)

Interesting...is this the only required field here?

Reply:
no... i guess is should use the ConnectionString.parse() second parameter (required fields) too.

---
In reply to: [157023183](https://github.com/Azure/azure-iot-sdk-node/pull/179#discussion_r157023183) [](ancestors = 157023183)